### ISSUE TITLE
When system time jumps, snapserver will send outdated chunks and client stops playing

### ISSUE THREAD

**Issue by TimoKersten** [badaix/snapcast#522]
### TLDR
Snapclient stops playing music shortly after system boot. In this error mode, the server sends chunks that are too old for the client to play. The situation occurs because I assume, a little while after boot timesyncd adjusts the system time forward and this adjustment is not reflected in the timestamps calculated in `ProcessStream::worker()`.

### Bug description:
When I boot my raspberry pi with snapserver and snapclient installed it is initially possible to play music for a few seconds. Afterward, the client won't play any music with the log looking like this: 

```Dec 30 18:44:23 wohnzimmer snapclient[503]: 2019-12-30 18-44-23 [Info] sleep > chunkDuration: 9844 > 26, time server now 1577731463873, chunk_->start: 1577731453149, chunks: 11, out: 119, needed: 30
Dec 30 18:44:23 wohnzimmer snapclient[503]: 2019-12-30 18-44-23 [Info] sleep > chunkDuration: 9818 > 26, time server now 1577731463874, chunk_->start: 1577731453175, chunks: 10, out: 119, needed: 30
Dec 30 18:44:23 wohnzimmer snapclient[503]: 2019-12-30 18-44-23 [Info] sleep > chunkDuration: 9792 > 26, time server now 1577731463874, chunk_->start: 1577731453201, chunks: 9, out: 119, needed: 30
Dec 30 18:44:23 wohnzimmer snapclient[503]: 2019-12-30 18-44-23 [Info] sleep > chunkDuration: 9766 > 26, time server now 1577731463874, chunk_->start: 1577731453227, chunks: 8, out: 119, needed: 30
Dec 30 18:44:23 wohnzimmer snapclient[503]: 2019-12-30 18-44-23 [Info] sleep > chunkDuration: 9740 > 26, time server now 1577731463874, chunk_->start: 1577731453254, chunks: 7, out: 119, needed: 30
Dec 30 18:44:23 wohnzimmer snapclient[503]: 2019-12-30 18-44-23 [Info] sleep > chunkDuration: 9714 > 26, time server now 1577731463874, chunk_->start: 1577731453280, chunks: 6, out: 119, needed: 30
```

Note that for all these chunks the server time is ahead of the chunk time by roughly 10 seconds. Thus, the client drops all these chunks and requests newer ones from the server. However, the server only sends chunks with too old time stamps. 

### Cause
A little while after boot timesyncd adjusts the system time forward and this adjustment is not reflected in the timestamps calculated in `ProcessStream::worker()`. `worker` gets the current system time as timestamp for chunks and then reads chunks from a pipe in a while loop. In each loop iteration the timestamp is moved forward by incrementing it by the time according to number of samples read (https://github.com/badaix/snapcast/blob/865701aa0fc7cea209e009cb6d1173a1f2f59ed4/server/streamreader/process_stream.cpp#L204).

As far as I have seen the adjusted system time is never transferred to the timestamp, so the server just keeps sending out chunks with old time stamps.

### Possible solutions
a) Use `std::chrono::steady_clock` instead of `std::chrono::system_clock`. @badaix have you experimented with stead_clock before? Are there any arguments to not use it?
b) Check whether timestamps and system clock diverged within the loop. Adjust timestamp in case divergence is too large. 

I implemented b) to check whether my bug analysis is correct: https://github.com/TimoKersten/snapcast/commit/437a5081f04a31ce3e814ed955877ce36224fd7f
It is only a makeshift solution but removed the bug. Personally, I would prefer to merge solution a) into this repository. @badaix any preferences?

**Comment by TimoKersten** [comment]
Ok, I just switched the used clock to steady_clock and it nicely removes the bug: https://github.com/TimoKersten/snapcast/commit/0876b67e0c237e6068897f157e23586457f919eb

Added bonus: When snapserver switches to steady clock and a snapclient is not yet updated it still works. I did see some initial silence, but the client did start playing audio after a few seconds.

I guess using solution a) is a good idea then.

**Comment by badaix** [comment]
Thanks, really nice bug report :)
Actually, I've never used steady_clock, I just switched once from `system_clock` to `steady_clock`, which resulted in an overflow somewhere in the chain.
I'm also aware of this "hick-up", when the server's ntp daemon does the initial time adjustment.
I didn't spent any effort in this, because the steady clock is perfectly suited for measuring time intervals, but might run actually faster/slower than the system clock, so that I am a bit afraid that the clients have to adjust permanently their clocks.
I will test the steady clock approach, maybe it can even be made configurable.

**Comment by badaix** [comment]
Thanks again! It's merged to master and will be part of v0.19
